contracts/cw3-fixed-multisig/src/contract.rs

```diff
@@ -399,5 +457,485 @@ fn calc_range_start(start_after: Option<HumanAddr>) -> Option<Vec<u8>> {
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // Proposal from voter works
+        let env = mock_env(VOTER3, &[]);
+        let res = handle(&mut deps, env, proposal).unwrap();
+
+        // Verify
+        assert_eq!(
+            res,
+            HandleResponse {
+                messages: vec![],
+                log: vec![
+                    log("action", "propose"),
+                    log("sender", VOTER3),
+                    log("proposal_id", 1),
+                ],
+                data: None
+            }
+        )
+    }
+
+    #[test]
+    fn test_vote_works() {
+        let mut deps = mock_dependencies(20, &[]);
+
+        let required_weight = 3;
+        let voting_period = Duration::Time(2000000);
+
+        let env = mock_env(OWNER, &[]);
+        setup_test_case(&mut deps, env.clone(), required_weight, voting_period).unwrap();
+
+        // Propose
+        let bank_msg = BankMsg::Send {
+            from_address: OWNER.into(),
+            to_address: SOMEBODY.into(),
+            amount: vec![coin(1, "BTC")],
+        };
+        let msgs = vec![CosmosMsg::Bank(bank_msg)];
+        let proposal = HandleMsg::Propose {
+            title: "Pay somebody".to_string(),
+            description: "Do I pay her?".to_string(),
+            msgs,
+            latest: None,
+        };
+        let res = handle(&mut deps, env.clone(), proposal).unwrap();
+
+        // Get the proposal id from the logs
+        let proposal_id: u64 = res.log[2].value.parse().unwrap();
+
+        // Owner cannot vote (again)
+        let vote = HandleMsg::Vote {
+            proposal_id,
+            vote: Vote::Yes,
+        };
+        let res = handle(&mut deps, env, vote.clone());
+
+        // Verify
+        assert!(res.is_err());
+        match res.unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(&msg, "Already voted on this proposal"),
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // Only voters can vote
+        let env = mock_env(SOMEBODY, &[]);
+        let res = handle(&mut deps, env, vote.clone());
+
+        // Verify
+        assert!(res.is_err());
+        match res.unwrap_err() {
+            StdError::Unauthorized { .. } => {}
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // But voter1 can
+        let env = mock_env(VOTER1, &[]);
+        let res = handle(&mut deps, env, vote.clone()).unwrap();
+
+        // Verify
+        assert_eq!(
+            res,
+            HandleResponse {
+                messages: vec![],
+                log: vec![
+                    log("action", "vote"),
+                    log("sender", VOTER1),
+                    log("proposal_id", proposal_id),
+                    log("status", "Open"),
+                ],
+                data: None
+            }
+        );
+
+        // TODO: No/Veto votes have no effect on the tally
```

Comment:
The only thing I would add here are the voting cases you already marked with TODO.

Other than that, looking great and happy to merge when the last points are finished